### PR TITLE
Annotations new features

### PR DESCRIPTION
**Pull request by Marcel-Mueck**
# What

The new update to the annotation pipeline includes all annotation tools used for the published version of deeprvat, as well as all processing steps needed to create ready-to-run annotation data for the published deeprvat version. 
annotation_rulegraph

Updates since last push:
- inclusion of AlphaMissense annotations as plugin in VEP
- changed DAG in order to optimize parallelisation
- Allele Frequency, MAF and AF_MB calculated from genotype file
- pipeline includes filtering step for variants based on distance to the nearest exon
- selection of required columns for deeprvat as well as renaming of columns to fit required input is now performed
- abSplice pipeline is now integrated into the DAG
- spliceAI scores are aggregated to spliceAI_delta_sores

# Testing
- since the annotation tools rely on pre-scored data, automatic tests via GitHub actions is not completely possible
- the annotation pipeline was instead tested locally using example data

### CODE REVIEW

**Comment by meyerkm** [comment]
@Marcel-Mueck There seems to be duplicate functions in annotations.py for the following functions: 
- deepripe_score_variant_onlyseq_all
- merge_abscore
- process_annotations

Also, looks like the Read-the-docs build is failing as the docstrings aren't in the correct Sphinx format. See here to adapt: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

File: deeprvat/annotations/annotations.py [hunk 1]
```diff
@@ -1,12 +1,13 @@
+import itertools
```
Comment (endast) [inline deeprvat/annotations/annotations.py:1]: This one is never used

File: deeprvat/annotations/annotations.py [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 import logging
 import os
 import pickle
 import random
 import sys
 import time
 from pathlib import Path
-from typing import Optional
-
+from typing import List, Optional
```
Comment (endast) [inline deeprvat/annotations/annotations.py:9]: This one is never used

**Review by endast — CHANGES_REQUESTED** [review]

File: pipelines/annotations.snakefile [hunk 3]
```diff
@@ -95,47 +96,218 @@ pvcf_blocks_df["chr_name"] = chr_mapping.loc[pvcf_blocks_df["Chromosome"].values
...
 block = pvcf_blocks_df["Block"]
 
+# init absplice
+absplice_repo_dir = Path(config["absplice_repo_dir"])
+n_cores_absplice = int(config.get("n_cores_absplice") or 4)
+ncores_merge_absplice = int(config.get("n_cores_merge_absplice") or 8)
+ncores_agg_absplice = int(config.get("ncores_agg_absplice") or 4)
+
+
+SNAKEMAKE_WORKDIR = config.get('absplice_tmp_dir') or  absplice_repo_dir /'example'/ 'workflow'
```
Comment (Marcel-Mueck) [inline pipelines/annotations.snakefile:106]: remove

File: pipelines/annotations.snakefile [hunk 4]
```diff
@@ -95,47 +96,218 @@ pvcf_blocks_df["chr_name"] = chr_mapping.loc[pvcf_blocks_df["Chromosome"].values
 chromosomes = pvcf_blocks_df["chr_name"]
 block = pvcf_blocks_df["Block"]
 
+# init absplice
+absplice_repo_dir = Path(config["absplice_repo_dir"])
```
Comment (Marcel-Mueck) [inline pipelines/annotations.snakefile:100]: remove

File: pipelines/annotations.snakefile [hunk 5]
```diff
@@ -95,47 +96,218 @@ pvcf_blocks_df["chr_name"] = chr_mapping.loc[pvcf_blocks_df["Chromosome"].values
...
+n_cores_absplice = int(config.get("n_cores_absplice") or 4)
+ncores_merge_absplice = int(config.get("n_cores_merge_absplice") or 8)
+ncores_agg_absplice = int(config.get("ncores_agg_absplice") or 4)
+
+
+SNAKEMAKE_WORKDIR = config.get('absplice_tmp_dir') or  absplice_repo_dir /'example'/ 'workflow'
+absplice_download_dir = config.get('absplice_download_dir') or  absplice_repo_dir /'example'/'data'/'resources'/'downloaded_files'
+absplice_output_dir = config.get('absplice_output_dir', anno_tmp_dir /'absplice')
+vcf_id = anno_tmp_dir / '{vcf_id}'
+vcf_dir = os.path.dirname(vcf_id)
```
Comment (Marcel-Mueck) [inline pipelines/annotations.snakefile:110]: why like this?

File: pipelines/config/deeprvat_annotation_config.yaml [hunk 6]
```diff
@@ -18,23 +19,17 @@ source_variant_dir : input_dir/vcf
...
 vep_repo_dir : repo_dir/ensembl-vep
-vep_plugin_repo : repo_dir/VEP_plugins
-variant_file_path : preprocessing_workdir/norm/variants/variants.tsv.gz
-pybedtools_tmp_path : output_dir/annotations/tmp/pybedtools
-n_jobs_deepripe : 32
-n_cores_merge_absplice : 32
-n_cores_absplice : 32
-deepsea_pca_pickle_filepath : annotations/deepSea_pca/pca.pkl
-deepsea_pca_n_components: 100
+preprocessing_workdir : preprocessing_workdir
```
Comment (Marcel-Mueck) [inline pipelines/config/deeprvat_annotation_config.yaml:29]: named accordingly?

File: pipelines/resources/absplice.yaml
Comment (Marcel-Mueck) [inline pipelines/resources/absplice.yaml:1]: pip freeze and lock versions

File: deeprvat/annotations/annotations.py [hunk 7]
```diff
@@ -16,10 +17,32 @@
 from joblib import Parallel, delayed
 from keras.models import load_model
 from sklearn.decomposition import PCA
-from tqdm import tqdm
+from tqdm import tqdm, trange
+import warnings
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply deeprvat/annotations/annotations.py:21]: rmvd

File: deeprvat/annotations/annotations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply deeprvat/annotations/annotations.py:1]: rmvd

File: deeprvat/annotations/annotations.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply deeprvat/annotations/annotations.py:9]: rmvd

File: pipelines/annotations.snakefile (hunk 5 above)
Comment (Marcel-Mueck) [reply pipelines/annotations.snakefile:110]: resolved

File: pipelines/annotations.snakefile [hunk 8]
```diff
@@ -95,47 +96,218 @@ pvcf_blocks_df["chr_name"] = chr_mapping.loc[pvcf_blocks_df["Chromosome"].values
...
+
+absplice_main_conf_path = deeprvat_parent_path / 'pipelines'/'resources'/"config_absplice.yaml"
+with open(absplice_main_conf_path, "r") as fd:
+    absplice_main_conf = yaml.safe_load(fd)
+
+
+include: Path('resources')/"absplice_download_Snakefile"
+include: Path('resources')/"absplice_splicing_pred_DNA_Snakefile"
+if absplice_main_conf['AbSplice_RNA'] == True:
+    include: deeprvat_parent_path / 'deeprvat' / 'pipelines'/'resources'/"absplice_splicing_pred_RNA_Snakefile"
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/annotations.snakefile:133]: done

File: pipelines/annotations.snakefile (hunk 4 above)
Comment (Marcel-Mueck) [reply pipelines/annotations.snakefile:100]: needed for default download dir

File: pipelines/config/deeprvat_annotation_config.yaml (hunk 6 above)
Comment (Marcel-Mueck) [reply pipelines/config/deeprvat_annotation_config.yaml:29]: no <parent directory> name defined in the preprocessing pipeline

File: pipelines/resources/absplice_download_Snakefile
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_download_Snakefile:1]: done

File: pipelines/resources/absplice_splicing_pred_DNA_Snakefile
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_DNA_Snakefile:1]: done

File: pipelines/resources/absplice_splicing_pred_RNA_Snakefile
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_RNA_Snakefile:1]: done

File: pipelines/resources/config_absplice.yaml [hunk 9]
```diff
@@ -0,0 +1,92 @@
+# ============================= AbSplice-DNA related ================================
+
+# specify genome version hg19 or hg38 
+genome: hg38
+extra_info_dna: False
+# store all the vcf file that you want to analyze to directory ../data/resources/vcf_files/
+#vcf: /omics/odcf/analysis/OE0540_projects/ukbb/exomes/all_200k/annotations/tmp/absplice_input/{vcf_id}
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/config_absplice.yaml:7]: rmvd

File: pipelines/resources/config_absplice.yaml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+# parameters for variant filtering
+variant_filtering:
+  minor_allele_frequency: 
+    apply_filter: False
+    filter_cutoff: 0.001
+  max_num_samples: 
+    apply_filter: False
+    filter_cutoff: 2
+
+# False by default, you can change this if you want to run on large datasets (see Readme)
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/config_absplice.yaml:18]: removed line

File: pipelines/resources/environment_gnomad_rocksdb.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  - python-rocksdb>=0.7
+  - rocksdb
+  - cython
+  - pandas
+  - pip
+  - snakemake>=6.1
+  - click
+  - wget
+  - bioconda::cyvcf2==0.30.16
+  # - bioconda::kipoiseq>=0.7
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/environment_gnomad_rocksdb.yaml:15]: rmvd unnecessary line

File: pipelines/resources/absplice_download_Snakefile [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        )
+    )
+
+
+rule all_download:
+    input:
+        list_outputs
+
+
+del splicemap5
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_download_Snakefile:147]: no idea honestly

File: pipelines/resources/absplice_splicing_pred_DNA_Snakefile [hunk 13]
```diff
@@ -0,0 +1,145 @@
+import os
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_DNA_Snakefile:1]: rmvd unnecessary imports

File: pipelines/resources/absplice_splicing_pred_DNA_Snakefile [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+import yaml
+from os import listdir
+from os.path import isfile, join
+from pathlib import Path
+
+genome = absplice_main_conf['genome']
+#vcf_dir = os.path.dirname(absplice_main_conf['vcf'])
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_DNA_Snakefile:8]: rmvd

File: pipelines/resources/absplice_splicing_pred_DNA_Snakefile [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    input:
+        mmsplice_splicemap = Path(absplice_output_dir) / config_pred['splicing_pred']['mmsplice_splicemap'],
+        spliceai = Path(absplice_output_dir) / config_pred['splicing_pred']['spliceai'],
+    params:
+        extra_info = absplice_main_conf['extra_info_dna']
+    conda:
+            "absplice"
+    output:
+        absplice_dna = absplice_output_dir / '{genome}' / 'dna' / '{vcf_id}_AbSplice_DNA.csv'
+        #absplice_dna = config_pred['splicing_pred']['absplice_dna']
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_DNA_Snakefile:131]: rmvd

File: pipelines/resources/absplice_splicing_pred_DNA_Snakefile [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    script:
+        "./absplice_dna.py"
+
+rule all_predict_dna:
+    input:
+        expand([absplice_output_dir / absplice_main_conf['genome'] / 'dna' / source_variant_file_pattern],zip, chr=chromosomes, block=block)
+        # expand( config_pred['splicing_pred']['absplice_dna'],
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_DNA_Snakefile:138]: rmvd

File: pipelines/resources/absplice_splicing_pred_RNA_Snakefile [hunk 17]
```diff
@@ -0,0 +1,98 @@
+import os
+import yaml
+from os import listdir
+from os.path import isfile, join
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_RNA_Snakefile:4]: rmvd unnecessary imports

File: pipelines/resources/absplice_splicing_pred_RNA_Snakefile [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+import pathlib
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_RNA_Snakefile:5]: changed - line is unnecessary though, I think, bc  Path is imported in main snakemake pipeline, from which this (sub-)pipeline is called

File: pipelines/resources/absplice_splicing_pred_RNA_Snakefile [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+include: "../DNA/rare_variants/Snakefile"
+include: "./count_table/Snakefile"
+include: "./outliers/Snakefile"
+
+genome = absplice_main_conf['genome']
+vcf_dir = os.path.dirname(absplice_main_conf['vcf'])
+vcf_ids = [file for file in listdir(vcf_dir) if '.tbi' not in file]
+
+def splicemap5(wildcards):
+    path = Path(config_download['splicemap_dir'])
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_splicing_pred_RNA_Snakefile:16]: added but see other commend

File: pipelines/resources/absplice.yaml
Comment (Marcel-Mueck) [reply pipelines/resources/absplice.yaml:1]: done

File: pipelines/resources/absplice_download_Snakefile [hunk 20]
```diff
@@ -0,0 +1,149 @@
+import os
+import yaml
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/absplice_download_Snakefile:2]: rmoved unnecessary imports

File: pipelines/resources/environment_spliceai_rocksdb.yaml [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+  - pip
+  - setuptools
+  - click
+  - pooch
+  - bioconda::snakemake>=6.1
+  - bioconda::kipoiseq
+  - bioconda::spliceai
+  - bioconda::cyvcf2
+  - pip:
+    - git+https://github.com/gagneurlab/spliceai_rocksdb.git@master
```
[the remark this answers is not in the document]
Comment (Marcel-Mueck) [reply pipelines/resources/environment_spliceai_rocksdb.yaml:22]: done

**Review by endast — APPROVED** [review]

**Pull request merged**